File: unicode/bench/bench.ts
// Run with node or bun after gen_input.py wrote out/input.txt.
import { readFileSync } from "node:fs";

const TEXT = readFileSync("out/input.txt", "utf8");

function utf8Cs(s: string): number {
  const b = new TextEncoder().encode(s);
  let h = 0;
  for (let i = 0; i < b.length; i++) h = (Math.imul(h, 31) + b[i]) >>> 0;
  return (h + b.length) >>> 0;
}

function graphemeCount(s: string): number {
  const seg = new Intl.Segmenter(undefined, { granularity: "grapheme" });
  let n = 0;
  for (const _ of seg.segment(s)) n++;
  return n;
}

let t0 = performance.now();
const nfc = TEXT.normalize("NFC");
let ms = performance.now() - t0;
console.log(`nfc\t${ms.toFixed(3)}\t${utf8Cs(nfc)}`);

t0 = performance.now();
const n = graphemeCount(TEXT);
ms = performance.now() - t0;
console.log(`graphemes\t${ms.toFixed(3)}\t${n}`);

File: unicode/bench/bench.py
import time
import unicodedata
from pathlib import Path

TEXT = (Path(__file__).resolve().parent / "out" / "input.txt").read_bytes().decode("utf-8")


def utf8_cs(s: str) -> int:
    b = s.encode("utf-8")
    h = 0
    for x in b:
        h = (h * 31 + x) & 0xFFFFFFFF
    return (h + len(b)) & 0xFFFFFFFF


t0 = time.perf_counter()
out = unicodedata.normalize("NFC", TEXT)
ms = (time.perf_counter() - t0) * 1000
print(f"nfc\t{ms:.3f}\t{utf8_cs(out)}")
